metricbeat/module/kubernetes/event/event_test.go

```diff
@@ -0,0 +1,117 @@
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/kubernetes"
+)
+
+func TestGenerateMapStrFromEvent(t *testing.T) {
+	labels := map[string]string{
+		"app.kubernetes.io/name":      "mysql",
+		"app.kubernetes.io/version":   "5.7.21",
+		"app.kubernetes.io/component": "database",
+	}
+
+	annotations := map[string]string{
+		"prometheus.io/path":   "/metrics",
+		"prometheus.io/port":   "9102",
+		"prometheus.io/scheme": "http",
+		"prometheus.io/scrape": "false",
+	}
+
+	mockEvent := v1.Event{
+		Metadata: &k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{
+			Labels:      labels,
+			Annotations: annotations,
+		},
+	}
+
+	expectedLabelsMapStrWithDot := common.MapStr{
+		"app": common.MapStr{
+			"kubernetes": common.MapStr{
+				"io/version":   "5.7.21",
+				"io/component": "database",
+				"io/name":      "mysql",
+			},
+		},
+	}
+	expectedLabelsMapStrWithDeDot := common.MapStr{
+		"app_kubernetes_io/name":      "mysql",
+		"app_kubernetes_io/version":   "5.7.21",
+		"app_kubernetes_io/component": "database",
+	}
+
+	expectedAnnotationsMapStrWithDot := common.MapStr{
+		"prometheus": common.MapStr{
+			"io/path":   "/metrics",
+			"io/port":   "9102",
+			"io/scheme": "http",
+			"io/scrape": "false",
+		},
+	}
+	expectedAnnotationsMapStrWithDeDot := common.MapStr{
+		"prometheus_io/path":   "/metrics",
+		"prometheus_io/port":   "9102",
+		"prometheus_io/scheme": "http",
+		"prometheus_io/scrape": "false",
+	}
+
+	watchOptions1 := kubernetes.WatchOptions{
+		LabelsDedot:      false,
+		AnnotationsDedot: false,
+	}
+	mapStrOutput1 := generateMapStrFromEvent(&mockEvent, watchOptions1)
+	metadata1 := mapStrOutput1["metadata"].(common.MapStr)
+	assert.Equal(t, expectedLabelsMapStrWithDot, metadata1["labels"])
+	assert.Equal(t, expectedAnnotationsMapStrWithDot, metadata1["annotations"])
+
+	watchOptions2 := kubernetes.WatchOptions{
+		LabelsDedot:      true,
+		AnnotationsDedot: false,
+	}
+	mapStrOutput2 := generateMapStrFromEvent(&mockEvent, watchOptions2)
+	metadata2 := mapStrOutput2["metadata"].(common.MapStr)
+	assert.Equal(t, expectedLabelsMapStrWithDeDot, metadata2["labels"])
+	assert.Equal(t, expectedAnnotationsMapStrWithDot, metadata2["annotations"])
+
+	watchOptions3 := kubernetes.WatchOptions{
+		LabelsDedot:      false,
+		AnnotationsDedot: true,
+	}
+	mapStrOutput3 := generateMapStrFromEvent(&mockEvent, watchOptions3)
+	metadata3 := mapStrOutput3["metadata"].(common.MapStr)
+	assert.Equal(t, expectedLabelsMapStrWithDot, metadata3["labels"])
+	assert.Equal(t, expectedAnnotationsMapStrWithDeDot, metadata3["annotations"])
+
+	watchOptions4 := kubernetes.WatchOptions{
+		LabelsDedot:      true,
+		AnnotationsDedot: true,
+	}
+	mapStrOutput4 := generateMapStrFromEvent(&mockEvent, watchOptions4)
+	metadata4 := mapStrOutput4["metadata"].(common.MapStr)
+	assert.Equal(t, expectedLabelsMapStrWithDeDot, metadata4["labels"])
+	assert.Equal(t, expectedAnnotationsMapStrWithDeDot, metadata4["annotations"])
+}
```

Comment:
These `generateMapStrFromEvent` tests should be separate tests, so it's immediately clear what is tested, which one failed + one can run a single test via CLI:

```
cases := map[string]struct{
  opts kubernetes.WatchOptions
  want/validate ...
}{
  "dedot all": {
    opts: kubernetes.WatchOptions{LabelsDot: true, AnnotationsDedot: true},
    wants: ...
  }
}

for name, test := range cases {
  test := test
  t.Run(name, func(t *testing.T) {
    evt := ... // no shared data between tests for isolation
    out := generateMapStrFromEvent(&evt, test.opts)
    assert.Equal(t, test.want, out)
  })
}
```

This get's you a test named `TestGenerateMapStrFromEvent/detot_all`. One can run single tests via `-run`, which uses regexes like: `go test -run 'GenerateMapStr/_all'. The `-run` flag splits names on `/` and compares names using the regex via `regex.MatchString` (sub-strings will match).

This way it is easier to extend the tests in the future as well.

Reply:
@urso Thanks for the review! I think I understood part of your comments. I refactored this test using a `struct` includes `config`, `mockEvent` and `expectedMetadat` please see https://github.com/elastic/beats/blob/43191e1b965cdf6fd1c8e1c9b8da93780b751460/metricbeat/module/kubernetes/event/event_test.go#L76-#L79 

And then I used a for loop to test through these several cases. https://github.com/elastic/beats/blob/43191e1b965cdf6fd1c8e1c9b8da93780b751460/metricbeat/module/kubernetes/event/event_test.go#L147 

But I dont quite undertand the `TestGenerateMapStrFromEvent/detot_all` part. Do I need to rename the test or this is just an example of using regex in `go test -run` command? Thanks again!

Reply:
The change already makes the test much more readable, thanks.

Having mulitple test cases, each one should have a name. Top-level tests names are the function name itself. For giving a sub-test a name one uses  `t.Run(name, func(t *testing.T){ ... })`.

The name can be constructed in different means, but should be reable, directly pointing developers to the actual test/config that failed. A common/simple pattern to give table-driven sub-tests names is using a map (which also randomized order of tests between runs):

```
cases := map[string]struct{
  ...
}{
  "test1-name": {
    ...
  },
  "test2-name": {
    ...
  },
  ...
}
for name, test := range cases {
  test := test // create local var in case someone decides to enable parallel test runs
  t.Run(name, func(t *testing.T) {
    // test body
  })
}
```

When running you tests with `-v` you will see the actual test names. If one sub-test fails, then the failure is logged with the sub-test only. The same if you use `t.Log(...)`.
Without proper isolation/naming it is not easy to see which test failed, especially with the for loop.

Once you have names for sub-tests, you can select unique tests to run via `go test -run ...`.

In my sample "detot_all" is the generated sub-test name for the sub-test. `TestGenerateMapStrFromEvent/detot_all` is the full name of the subtest.

Reply:
@urso Perfect!! Thank you so much for all the detailed explanation. I just pushed a new commit with adding names for each test case. Thanks again!